File: JavaScript solutions/278. First Bad Version.js
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function (isBadVersion) {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */
  return function (n) {
    let f = 1;
    let l = n;
    while (f <= l) {
      let m = Math.floor((f + l) / 2);
      if (isBadVersion(m)) l = m - 1;
      else f = m + 1;
    }
    return f;
  };
};

File: C++ solutions/278. First Bad Version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        long long f=1,l=n;
        long long m=0;
        while(f<=l){
            m=(f+l)/2;
            if(isBadVersion(m)==true) l=m-1;
            else if(isBadVersion(m)==false) f=m+1;
        }
        return f;
    }
};